src/Compilers/CSharp/Test/Semantic/Semantics/GenericConstraintsTests.cs

```diff
@@ -2674,5 +2904,103 @@ public unsafe class C<U> where U : unmanaged
                 //     public void M1<T>() where T : T* { }
                 Diagnostic(ErrorCode.ERR_BadConstraintType, "T*").WithLocation(4, 35));
         }
+
+        [Fact]
+        public void UnmanagedConstraints_EnumWithUnmanaged()
+        {
+            Action<ModuleSymbol> validator = module =>
+            {
+                var typeParameter = module.GlobalNamespace.GetTypeMember("Test").TypeParameters.Single();
+
+                Assert.True(typeParameter.HasUnmanagedTypeConstraint);
+                Assert.True(typeParameter.HasValueTypeConstraint);
+                Assert.False(typeParameter.HasReferenceTypeConstraint);
+                Assert.False(typeParameter.HasConstructorConstraint);
+
+                Assert.Equal("Enum", typeParameter.ConstraintTypes().Single().Name);
+
+                Assert.True(typeParameter.IsValueType);
+                Assert.False(typeParameter.IsReferenceType);
+            };
+
+            CompileAndVerify(
+                source: "public class Test<T> where T : unmanaged, System.Enum {}",
+                sourceSymbolValidator: validator,
+                symbolValidator: validator);
+        }
+
+        [Fact]
+        public void UnmanagedConstraints_NestedInGenericType()
+        {
+            var code = @"
+public class Wrapper<T>
+{
+    public enum E
+    {
+    }
+
+    public struct S
+    {
+    }
+}
+public class Test
+{
+    void IsUnmanaged<T>() where T : unmanaged { }
+    void IsEnum<T>() where T : System.Enum { }
+    void IsStruct<T>() where T : struct { }
+    void IsNew<T>() where T : new() { }
+    
+
+    void User()
+    {
+        IsUnmanaged<Wrapper<int>.E>();
+        IsEnum<Wrapper<int>.E>();
+        IsStruct<Wrapper<int>.E>();
+        IsNew<Wrapper<int>.E>();
+
+        IsUnmanaged<Wrapper<int>.S>();          // Invalid
+        IsEnum<Wrapper<int>.S>();               // Invalid
+        IsStruct<Wrapper<int>.S>();
+        IsNew<Wrapper<int>.S>();
+    }
```

Comment:
Consider testing with reference type argument as well: `Wrapper<string>.E` and `Wrapper<string>.S` perhaps.

Reply:
Added them, although in this particular scenario, the type of `T` does not make a difference. We treat all generic types here the same.

---
In reply to: [177223597](https://github.com/dotnet/roslyn/pull/25680#discussion_r177223597) [](ancestors = 177223597)